libsolidity/formal/VariableUsage.cpp

```diff
@@ -30,8 +30,25 @@ void VariableUsage::endVisit(Identifier const& _identifier)
 	Declaration const* declaration = _identifier.annotation().referencedDeclaration;
 	solAssert(declaration, "");
 	if (VariableDeclaration const* varDecl = dynamic_cast<VariableDeclaration const*>(declaration))
-		if (_identifier.annotation().lValueRequested)
+	{
+		bool isLocal = varDecl->isLocalVariable();
+		bool inOuterFunction = false;
+		if (isLocal)
+		{
+			solAssert(m_outerFunction, "");
+			auto const& localVars = m_outerFunction->localVariables();
+			inOuterFunction = find(localVars.begin(), localVars.end(), varDecl) != localVars.end();
+			auto const& retVars = m_outerFunction->returnParameters();
+			inOuterFunction |= find_if(retVars.begin(), retVars.end(), [&] (auto var) { return var.get() == varDecl; }) != retVars.end();
+			auto const& paramVars = m_outerFunction->parameters();
+			inOuterFunction |= find_if(paramVars.begin(), paramVars.end(), [&] (auto var) { return var.get() == varDecl; }) != paramVars.end();
```

Comment:
@chriseth any better way to do this check?

Reply:
You could use the `scope` function on VariableDeclaration. It might also return a `block` but perhaps a variant of that function could be introduced that goes all the way up to the function?